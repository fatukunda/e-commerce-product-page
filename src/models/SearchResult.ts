export interface ISearchItem {
    price: number,
    image: string,
    vendor: string,
    title: string,
    id: string,
    link: string
}