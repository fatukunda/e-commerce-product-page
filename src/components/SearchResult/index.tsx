import { FunctionComponent, useState, useEffect } from "react";
import Dropdown, { IOption } from "../Dropdown";
import Loader from "../Loader";

import ItemCard from "../ItemCard";

import { useAppSelector } from "src/store/hooks";
import { ISearchItem } from "src/models/SearchResult";
import Menu from "../Menu";
import { sort } from "src/services/searchService";
import { useAppDispatch } from "src/store/hooks";
import { setProducts, setIsSearching } from "src/store/slices/searchSlice";

const SearchResult: FunctionComponent = (): JSX.Element => {
  const searchedItems: ISearchItem[] = useAppSelector(
    (state) => state.search.items
  );
  const dispatch = useAppDispatch();
  const isSearching: boolean = useAppSelector(
    (state) => state.search.isSearching
  );
  const searchOptions: IOption[] = [
    {
      id: 1,
      name: "Random",
    },
    {
      id: 2,
      name: "Ascending",
    },
    {
      id: 3,
      name: "Descending",
    },
    {
      id: 4,
      name: "Price Low to High",
    },
    {
      id: 5,
      name: "Price High to Low",
    },
  ];
  const [selectedOption, setSelectedOption] = useState({
    id: 1,
    name: "Random",
  });
  const [isMenuOpen, setIsMenuOpen] = useState(false);
  const selectedOptionHandler = (option: IOption) => {
    setSelectedOption(option);
    setIsMenuOpen(false);
  };
  const closeMenuHandler = () => {
    setIsMenuOpen(false);
  };
  const toggleOpenHandler = () => {
    setIsMenuOpen(!isMenuOpen);
  };

  useEffect(() => {
    const search = async () => {
      let sortOption = "";
      if (selectedOption.name === "Ascending") {
        sortOption = "asc";
      } else if (selectedOption.name === "Descending") {
        sortOption = "desc";
      } else if (selectedOption.name === "Price Low to High") {
        sortOption = "price-low-high";
      } else if (selectedOption.name === "Price High to Low") {
        sortOption = "price-high-low";
      } else {
        sortOption = "Random";
      }
      dispatch(setIsSearching(true));
      const res = await sort(sortOption);
      dispatch(setIsSearching(false));
      dispatch(setProducts(res));
    };
    search();
  }, [selectedOption, dispatch]);

  return (
    <Menu classList="h-2/3 left-1/2 z-10 w-2/3  -translate-x-1/2 rounded-lg bg-white shadow-2xl shadow-tertiary/70 md:right-0 md:top-20  lg:top-24 lg:right-0 ">
      <div className="flex border-secondary-dark-dark block border-b border-solid p-6 font-bold items-center justify-between">
        <span className="block font-bold ">Your search results</span>
        <div>
          <Dropdown
            options={searchOptions}
            selectedItem={selectedOption}
            change={selectedOptionHandler}
            close={closeMenuHandler}
            isOpen={isMenuOpen}
            toggleOpen={toggleOpenHandler}
          />
        </div>
      </div>

      {!isSearching ? (
        <div className="p-4 flex gap-2 flex-wrap justify-center">
          {searchedItems?.length ? (
            searchedItems.map((item) => <ItemCard key={item.id} item={item} />)
          ) : (
            <div className="p-4 mt-36">
              <span className="inline-block w-full text-center font-bold text-tertiary-light">
                Oh no! No result for that search
              </span>
            </div>
          )}
        </div>
      ) : (
        <Loader />
      )}
    </Menu>
  );
};

export default SearchResult;
