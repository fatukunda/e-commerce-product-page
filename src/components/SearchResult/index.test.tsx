import { render, cleanup, fireEvent, getByTitle } from "@testing-library/react";
import renderer from "react-test-renderer";
import { Provider as ReduxProvider } from "react-redux";

import SearchResult from ".";
import mockStore from "src/components/mocks/store";

const renderSearchResult = () =>
  render(
    <ReduxProvider store={mockStore()}>
      <SearchResult />
    </ReduxProvider>
  );

describe("Tests for <SearchResult/>", () => {
  afterEach(cleanup);
  test("Search Result: Snapshot", () => {
    const component = renderer.create(
      <ReduxProvider store={mockStore()}>
        <SearchResult />
      </ReduxProvider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("Renders without crashing", () => {
    renderSearchResult();
  });
  it('Filters results in ascending order', () => {
    const { getByText, getAllByTitle, getByTitle } = renderSearchResult();
    const filterBtn = getByTitle('select-filter-btn')
    expect(getByText('Random')).toBeInTheDocument()
    fireEvent.click(filterBtn)
    const options = getAllByTitle('option')
    expect(options).toHaveLength(5)
    fireEvent.click(options[1])
    expect(filterBtn).toHaveTextContent('Ascending')
  });
  it('Filters results in descending order', () => {
    const { getByText, getAllByTitle, getByTitle } = renderSearchResult();
    const filterBtn = getByTitle('select-filter-btn')
    expect(getByText('Random')).toBeInTheDocument()
    fireEvent.click(filterBtn)
    const options = getAllByTitle('option')
    fireEvent.click(options[2])
    expect(filterBtn).toHaveTextContent('Descending')
  });
  it('Filters results in descending order', () => {
    const { getByText, getAllByTitle, getByTitle } = renderSearchResult();
    const filterBtn = getByTitle('select-filter-btn')
    expect(getByText('Random')).toBeInTheDocument()
    fireEvent.click(filterBtn)
    const options = getAllByTitle('option')
    fireEvent.click(options[3])
    expect(filterBtn).toHaveTextContent('Price Low to High')
  });
  it('Filters results in descending order', () => {
    const { getByText, getAllByTitle, getByTitle } = renderSearchResult();
    const filterBtn = getByTitle('select-filter-btn')
    expect(getByText('Random')).toBeInTheDocument()
    fireEvent.click(filterBtn)
    const options = getAllByTitle('option')
    fireEvent.click(options[4])
    expect(filterBtn).toHaveTextContent('Price High to Low')
  });
});
