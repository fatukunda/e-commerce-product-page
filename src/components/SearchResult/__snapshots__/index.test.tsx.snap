// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <SearchResult/> Search Result: Snapshot 1`] = `
<div
  className="absolute top-[70px] overflow-auto h-2/3 left-1/2 z-10 w-2/3  -translate-x-1/2 rounded-lg bg-white shadow-2xl shadow-tertiary/70 md:right-0 md:top-20  lg:top-24 lg:right-0 "
>
  <div
    className="flex border-secondary-dark-dark block border-b border-solid p-6 font-bold items-center justify-between"
  >
    <span
      className="block font-bold "
    >
      Your search results
    </span>
    <div>
      <div
        className="flex flex-col w-48"
      >
        <button
          className="flex bg-white items-center justify-between 
  btn-sm w-48 text-xs hover:bg-gray-100 text-tertiary-light font-semibold 
  py-2 px-4  border border-tertiary-light rounded shadow"
          onClick={[Function]}
          title="select-filter-btn"
        >
          <span
            title="selected-item"
          >
            Random
          </span>
          <svg
            style={
              Object {
                "height": "24px",
                "width": "24px",
              }
            }
            viewBox="0 0 24 24"
          >
            <path
              d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
  <div
    className="p-4 flex gap-2 flex-wrap justify-center"
  >
    <div
      className="w-48 bg-white rounded-lg shadow-md dark:border-gray-700"
    >
      <img
        alt="test product title"
        className="p-8 rounded-t-lg"
        src="https://image-link"
      />
      <div
        className="px-5 pb-5"
      >
        <h5
          className="text-xs text-tertiary-light"
        >
          test product title
        </h5>
        <div
          className="flex justify-between items-center"
        >
          <span
            className="text-l pt-2 font-bold text-gray-900"
          >
            â‚¬
            20
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
